photon-api/src/main/scala/com/linkedin/photon/ml/model/RandomEffectModel.scala

```diff
@@ -233,67 +236,44 @@ object RandomEffectModel {
+  protected def determineModelType(models: DataFrame): TaskType = {
 
-    val modelTypes = modelsRDD.values.map(_.modelType).distinct().collect()
+    val modelTypes = models.select(GeneralizedLinearModel.MODEL_TYPE).head(1)
 
     require(
       modelTypes.length == 1,
-      s"${modelsRDD.name} has multiple model types:\n${modelTypes.mkString(", ")}")
+      s"models has multiple model types:\n${modelTypes.mkString(", ")}")
 
-    modelTypes.head
+    TaskType.withName(modelTypes(0).getString(0))
   }
 
   /**
    * Compute the scores for a dataset, using random effect models.
    *
-   * @param dataPoints The dataset to score
-   * @param modelsRDD The individual random effect models to use for scoring
+   * @param dataset The dataset to score
+   * @param models The individual random effect models to use for scoring
    * @param randomEffectType The random effect type
    * @param featureShardId The feature shard id
    * @return The scores
    */
-  private def score[T, V](
-      dataPoints: RDD[(UniqueSampleId, GameDatum)],
-      modelsRDD: RDD[(REId, GeneralizedLinearModel)],
+  private def score (
+      dataset: DataFrame,
+      models: DataFrame,
       randomEffectType: REType,
-      featureShardId: FeatureShardId,
-      toScore: (GameDatum, Double) => T,
-      toResult: RDD[(UniqueSampleId, T)] => V): V = {
-
-    val hashPartitioner = new HashPartitioner(dataPoints.getNumPartitions)
-
-    /*
-     * We perform a replicated partitioned hash join here under the assumption that we can fit the per partition
-     * random effect models in memory. We first partition both relations using the same partitioner and then zip them.
-     * This ensures that the same keys from both relations go in the same partition. Given above, we can now perform the
-     * join by doing the following operations per partition:
-     *   1. Load the random effect models in memory
-     *   2. Iterate over the data points
-     *   3. For each data point, look up the corresponding random effect model in the in memory map and score
-     */
-    val scores = dataPoints
-      .map { case (uniqueId, gameDatum) =>
-        (gameDatum.idTagToValueMap(randomEffectType), (uniqueId, gameDatum))
-      }
-      .partitionBy(hashPartitioner)
-      .zipPartitions(modelsRDD.partitionBy(hashPartitioner)) { (dataIt, modelIt) =>
-
-        val lookupTable = modelIt.toMap
+      featureShardId: FeatureShardId): DataFrame = {
 
-        dataIt.map { case (id, (uid, datum)) =>
-          val score = lookupTable
-            .get(id)
-            .map(_.computeScore(datum.featureShardContainer(featureShardId)))
-            .getOrElse(0.0)
+    val scores: DataFrame = dataset
+      .join(models, randomEffectType)
+      .withColumn(DataConst.SCORE, GeneralizedLinearModel.scoreUdf(col(DataConst.COEFFICIENTS), col(featureShardId)))
+      .select(Constants.UNIQUE_SAMPLE_ID, DataConst.SCORE)
```

Comment:
Should not drop columns for model scoring - this is a result that could be passed to the user if they run Photon ML from the command line. The user should decide which columns to keep and which to merge/drop.

Reply:
Hmmm, I removed line 271; am wondering whether we should rename the score field name to ${randomEffectType}-score?